searx/network/context.py

```diff
@@ -0,0 +1,387 @@
+        self._http_client: Optional[ABCHTTPClient] = None
+        self._start_time: float = start_time or default_timer()
+        self._http_time: float = 0.0
+        self._timeout: Optional[float] = timeout
+
+    @abstractmethod
+    def call(self, func: Callable[P, R], *args: P.args, **kwargs: P.kwargs) -> R:
+        """Call func within the network context.
+
+        The retry policy might call func multiple times.
+
+        Within the function self.get_http_client() returns an HTTP client to use.
+
+        The retry policy might send multiple times the same HTTP request
+        until it works or the retry count falls to zero.
+        """
+
+    @final
+    def request(self, *args, **kwargs):
+        """Convenient method to wrap a call to request inside the call method.
+
+        Use a new HTTP client to wrap a call to the request method using self.call
+        """
+        self._set_http_client()
+        return self.call(self._get_http_client().request, *args, **kwargs)
+
+    @final
+    def stream(self, *args, **kwargs):
+        """Convenient method to wrap a call to stream inside the call method.
+
+        Use a new HTTP client to wrap a call to the stream method using self.call
+        """
+        self._set_http_client()
+        return self.call(self._get_http_client().stream, *args, **kwargs)
+
+    @final
+    def get_http_runtime(self) -> Optional[float]:
+        """Return the amount of time spent on HTTP requests"""
+        return self._http_time
+
+    @final
+    def get_remaining_time(self, _override_timeout: Optional[float] = None) -> float:
+        """Return the remaining time for the context.
+
+        _override_timeout is not intended to be used outside this module.
+        """
+        timeout = _override_timeout or self._timeout or DEFAULT_TIMEOUT
+        timeout += 0.2  # overhead
+        timeout -= default_timer() - self._start_time
+        return timeout
+
+    @final
+    def _get_http_client(self) -> ABCHTTPClient:
+        """Return the HTTP client to use for this context."""
+        if self._http_client is None:
+            raise ValueError("HTTP client has not been set")
+        return self._http_client
+
+    @final
+    def _set_http_client(self):
+        """Ask the NetworkContext to use another HTTP client using the factory.
+
+        Use the method _get_new_client_from_factory() to call the factory,
+        so the NetworkContext implementations can wrap the ABCHTTPClient.
+        """
+        self._http_client = self._get_new_client_from_factory()
+
+    @final
+    def _reset_http_client(self):
+        self._http_client = None
+
+    def _get_new_client_from_factory(self):
+        return self._http_client_factory()
+
+    @contextmanager
+    def _record_http_time(self):
+        """This decorator records the code's runtime and adds it to self.total_time"""
+        time_before_request = default_timer()
+        try:
+            yield
```

Comment:
Could you please explain why that is needed/what it does? I don't really get that.

Reply:
This follow the pattern from documentation: https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager

The following code 
```python
with _record_http_time():
  do_something()
```

is equivalent to:

```python
time_before_request = default_timer()
do_something()
self._http_time += default_timer() - time_before_request
```

even if `do_something()` raises an exception.